server/src/main/java/org/opensearch/search/pipeline/Pipeline.java

```diff
@@ -41,33 +45,55 @@ class Pipeline {
 
     // TODO: Refactor org.opensearch.ingest.CompoundProcessor to implement our generic Processor interface
     // Then these can be CompoundProcessors instead of lists.
-    private final List<SearchRequestProcessor> searchRequestProcessors;
-    private final List<SearchResponseProcessor> searchResponseProcessors;
+    private final List<ProcessorWithMetrics<SearchRequestProcessor>> searchRequestProcessors;
+    private final List<ProcessorWithMetrics<SearchResponseProcessor>> searchResponseProcessors;
 
     private final NamedWriteableRegistry namedWriteableRegistry;
+    private final SearchPipelineMetrics totalRequestMetrics;
+    private final SearchPipelineMetrics totalResponseMetrics;
+    private final SearchPipelineMetrics pipelineRequestMetrics = new SearchPipelineMetrics();
+    private final SearchPipelineMetrics pipelineResponseMetrics = new SearchPipelineMetrics();
```

Comment:
Metrics logic is really tangled in with this class. 
Personally I would just abstract it out to keep `Pipeline.java` simple.

Reply:
Whatever class invokes a pipeline and the individual processors needs to hold the references to the various metrics (in order to invoke `before`, `after`, and `failed` on the metrics).

I guess I could subclass `Pipeline`, creating a `PipelineWithMetrics`. The base class could define empty methods `beforeTransformRequest`, `afterTransformRequest`, `beforeRequestProcessor(String processorKey)`, `afterRequestProcessor(String processorKey)`, etc. The subclass would hold the metrics instances and delegate to the various metrics instances.

Is that what you had in mind?

Reply:
Yup I like that, mostly what I was looking for. 
Also I would take it further, defining `request`/`response` processors and letting `Pipeline.java` know about it is not really needed. I would rather pass the Processor and its got its Type as one of the parameters.

Reply:
I took the subclassing approach. I was worried that it would be hairy, but after taking a shot at it, the separation between "transforming stuff" and "tracking / managing metrics" looks (reasonably) clean.

Thanks for the suggestion!